resolwe/flow/models/utils.py

```diff
@@ -339,45 +339,74 @@ def hydrate_input_uploads(input_, input_schema, hydrate_values=True):
+        total_size = 0
+        for dirpath, _, filenames in os.walk(path):
+            for file_name in filenames:
+                file_path = os.path.join(dirpath, file_name)
+                if not os.path.isfile(file_path):  # Skip (broken) links
+                    continue
+                total_size += os.path.getsize(file_path)
+        return total_size
+
+    def get_refs_size(obj, obj_path):
+        """Caculate size of all references."""
+        total_size = 0
+        for ref in obj.get('refs', []):
+            ref_path = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.pk), ref)
+            if os.path.isfile(ref_path):
+                total_size += os.path.getsize(ref_path)
+            elif os.path.isdir(ref_path):
+                total_size += get_dir_size(ref_path)
+            elif os.path.exists(ref_path):
+                raise ValidationError("Reference is not a file or directory ({})".format(ref_path))
+            else:
+                raise ValidationError("Reference does not exist ({})".format(ref_path))
+
+            # It is a common case that ``obj['file']`` is also contained in
+            # one of obj['ref']. In that case, we need to make sure that it's
+            # size is not counted twice:
+            if ref_path in obj_path:
+                total_size -= obj['size']
+                assert total_size > 0
+
+        return total_size
+
     def add_file_size(obj):
         """Add file size to the basic:file field."""
-        if data.status in [Data.STATUS_DONE, Data.STATUS_ERROR] and 'size' in obj:
+        if data.status in [Data.STATUS_DONE, Data.STATUS_ERROR] and 'size' in obj and not force:
             return
 
         path = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.pk), obj['file'])
         if not os.path.isfile(path):
             raise ValidationError("Referenced file does not exist ({})".format(path))
 
         obj['size'] = os.path.getsize(path)
-
-    def get_dir_size(path):
-        """Get directory size."""
-        total_size = 0
-        for dirpath, _, filenames in os.walk(path):
-            for file_name in filenames:
-                file_path = os.path.join(dirpath, file_name)
-                total_size += os.path.getsize(file_path)
-        return total_size
+        obj['total_size'] = obj['size'] + get_refs_size(obj, path)
 
     def add_dir_size(obj):
         """Add directory size to the basic:dir field."""
-        if data.status in [Data.STATUS_DONE, Data.STATUS_ERROR] and 'size' in obj:
+        if data.status in [Data.STATUS_DONE, Data.STATUS_ERROR] and 'size' in obj and not force:
             return
 
         path = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.pk), obj['dir'])
         if not os.path.isdir(path):
             raise ValidationError("Referenced dir does not exist ({})".format(path))
 
         obj['size'] = get_dir_size(path)
+        obj['total_size'] = obj['size'] + get_refs_size(obj, path)
```

Comment:
Other option would be to omit `obj['size']` here and calculate size of whole reference.

Reply:
Yes, i was thinking about that. I assumed that `total_size` will be used more often than "ref_size", so it makes more sense to compute it directly, instead of having to sum up `size` and "ref_size" to get it.